test/util/regtest.js

```diff
@@ -0,0 +1,317 @@
+    maxOutbound: 1,
+    seeds: [],
+    nodes: [`127.0.0.1:${options.ports.full.p2p}`],
+    memory: false,
+    plugins: [require('../../lib/wallet/plugin')],
+    env: {
+      'BCOIN_WALLET_HTTP_PORT': (options.ports.spv.wallet).toString()
+    },
+    logLevel: options.logLevel
+  });
+
+  await node.ensure();
+  await node.open();
+  await node.connect();
+  await node.startSync();
+  return node;
+}
+
+async function initNodeClient(options) {
+  const nclient = new NodeClient({
+    network: 'regtest',
+    port: options.ports.node,
+    apiKey: 'foo'
+  });
+  await nclient.open();
+  return nclient;
+}
+
+async function initWalletClient(options) {
+  const wclient = new WalletClient({
+    network: 'regtest',
+    port: options.ports.wallet,
+    apiKey: 'foo'
+  });
+  await wclient.open();
+  return wclient;
+}
+
+async function initWallet(wclient) {
+  const winfo = await wclient.createWallet('test');
+  assert.strictEqual(winfo.id, 'test');
+  const wallet = wclient.wallet('test', winfo.token);
+  await wallet.open();
+
+  // A lookahead is increased to avoid timing issues with bloom
+  // filters not being loaded in time and transactions not being
+  // broadcast to spv node wallets.
+  const info = await wallet.createAccount('blue', {
+    witness: true,
+    lookahead: 40
+  });
+  assert(info.initialized);
+  assert.strictEqual(info.name, 'blue');
+  assert.strictEqual(info.accountIndex, 1);
+  assert.strictEqual(info.m, 1);
+  assert.strictEqual(info.n, 1);
+
+  return wallet;
+}
+
+async function generateBlocks(count, nclient, coinbase) {
+  return await nclient.execute('generatetoaddress', [count, coinbase]);
+}
+
+async function generateRollback(depth, nclient) {
+  const invalidated = [];
+
+  for (let i = 0; i < depth; i++) {
+    const hash = await nclient.execute('getbestblockhash');
+    invalidated.push(hash);
+    await nclient.execute('invalidateblock', [hash]);
+  }
+
+  return invalidated;
+}
+
+async function generateReorg(depth, nclient, wclient, coinbase) {
+  const blockInterval = 600;
+
+  const invalidated = [];
+  let lastTime = null;
+
+  // Invalidate blocks to the depth.
+  for (let i = 0; i < depth; i++) {
+    const hash = await nclient.execute('getbestblockhash');
+    invalidated.push(hash);
+
+    // Get the time for the block before it's removed.
+    const lastBlock = await nclient.execute('getblock', [hash]);
+    lastTime = lastBlock.time;
+
+    await nclient.execute('invalidateblock', [hash]);
+  }
+
+  // Increase time so that blocks do not have
+  // the same time stamp as before.
+  lastTime += 10000;
+
+  // TODO remove
+  await sleep(1000);
+
+  // Mature coinbase transactions past depth
+  await generateBlocks(depth, nclient, coinbase);
+
+  const txids = await wclient.execute('resendwallettransactions');
+
+  const validated = [];
+
+  // Add new blocks back to the same height plus two
+  // so that it becomes the chain with the most work.
+  for (let c = 0; c < depth + 2; c++) {
```

Comment:
To reviewers: This is where it's necessary to add two blocks because of an issue with SPV reorganizations: https://github.com/bcoin-org/bcoin/issues/655

Reply:
The issue has been identified for this at https://github.com/bcoin-org/bcoin/pull/755